# Pandas Data Manipulation Exercise

## Objective

This Exercise will help you gain hands-on experience with key Pandas operations, including data creation, exploration, filtering, slicing, merging, concatenating, and grouping data. Each task aims to build foundational skills in manipulating data using Pandas.

## Instructions

Complete each task sequentially, answering all questions.
Add comments to explain your code where necessary.
Submit your work as a Jupyter Notebook (.ipynb) or a Python script (.py).

## Dataset Setup
#### **Create a DataFrame df_sales using the following data:** ####
                
| Order_ID | Product     | Category    | Quantity | Price | Store_Location |
|----------|-------------|-------------|----------|-------|----------------|
| 1        | Laptop      | Electronics | 3        | 800   | New York       |
| 2        | Headphones  | Electronics | 5        | 150   | San Francisco  |
| 3        | Chair       | Furniture   | 10       | 85    | New York       |
| 4        | Desk        | Furniture   | 2        | 200   | Chicago        |
| 5        | Monitor     | Electronics | 4        | 300   | San Francisco  |
| 6        | Lamp        | Furniture   | 7        | 40    | New York       |
| 7        | Smartphone  | Electronics | 6        | 600   | Chicago        |
| 8        | Sofa        | Furniture   | 1        | 500   | San Francisco  |


In [1]:


import pandas as pd

data = {
    'Order_ID': [1, 2, 3, 4, 5, 6, 7, 8],
    'Product': ['Laptop', 'Headphones', 'Chair', 'Desk', 'Monitor', 'Lamp', 'Smartphone', 'Sofa'],
    'Category': ['Electronics', 'Electronics', 'Furniture', 'Furniture', 'Electronics', 'Furniture', 'Electronics', 'Furniture'],
    'Quantity': [3, 5, 10, 2, 4, 7, 6, 1],
    'Price': [800, 150, 85, 200, 300, 40, 600, 500],
    'Store_Location': ['New York', 'San Francisco', 'New York', 'Chicago', 'San Francisco', 'New York', 'Chicago', 'San Francisco']
}

df_sales = pd.DataFrame(data)



In [2]:
df_sales

Order_ID     Product     Category  Quantity  Price Store_Location
0         1      Laptop  Electronics         3    800       New York
1         2  Headphones  Electronics         5    150  San Francisco
2         3       Chair    Furniture        10     85       New York
3         4        Desk    Furniture         2    200        Chicago
4         5     Monitor  Electronics         4    300  San Francisco
5         6        Lamp    Furniture         7     40       New York
6         7  Smartphone  Electronics         6    600        Chicago
7         8        Sofa    Furniture         1    500  San Francisco

## Tasks and Questions

#### Task 1: Basic DataFrame Information

1.Display the first 5 rows and last 5 rows of df_sales using .head() and .tail().

**Question:** What are the products listed in the first and last rows?

In [3]:

df_sales.head()

Order_ID     Product     Category  Quantity  Price Store_Location
0         1      Laptop  Electronics         3    800       New York
1         2  Headphones  Electronics         5    150  San Francisco
2         3       Chair    Furniture        10     85       New York
3         4        Desk    Furniture         2    200        Chicago
4         5     Monitor  Electronics         4    300  San Francisco

**Answer:**

- The product listed in the first row is `Laptop`.
- The product listed in the last row is `Monitor`.

### **Data Structure**

Use .size, .shape, and .ndim to find the total number of elements, the number of rows and columns, and the dimensions of df_sales.

**Question:** How many rows and columns are in df_sales?



In [4]:

total_elements = df_sales.size


shape = df_sales.shape

dimensions = df_sales.ndim

print("Total number of elements:", total_elements)
print("Number of rows and columns:", shape)
print("Number of dimensions:", dimensions)


Total number of elements: 48
Number of rows and columns: (8, 6)
Number of dimensions: 2


`Answer:`

- Total number of elements: `48`
- Number of rows and columns: `(8, 6)`
- Number of dimensions: `2`


#### Column Information
Use .columns and .dtypes to check the column names and data types.

**Question:** Which columns contain numeric data?

In [5]:

columns = df_sales.columns

data_types = df_sales.dtypes

print("Columns:", columns)
print("\nData types of columns:")
print(data_types)


Columns: Index(['Order_ID', 'Product', 'Category', 'Quantity', 'Price',
       'Store_Location'],
      dtype='object')

Data types of columns:
Order_ID           int64
Product           object
Category          object
Quantity           int64
Price              int64
Store_Location    object
dtype: object


`Answer:`
The columns that contain numeric data are:

- Order_ID (int64)
- Quantity (int64)
- Price (int64)

## Task 2: Descriptive Statistics
#### Summary Statistics

Use .describe() to view summary statistics for numeric columns.

**Question:** What is the average quantity sold? What is the maximum price?


In [6]:
# CODE HERE
summary_statistics = df_sales.describe()

print(summary_statistics)

average_quantity = summary_statistics.loc["mean", "Quantity"]
maximum_price = summary_statistics.loc["max", "Price"]

print("\nAverage quantity sold:", average_quantity)
print("Maximum price:", maximum_price)


       Order_ID   Quantity       Price
count   8.00000   8.000000    8.000000
mean    4.50000   4.750000  334.375000
std     2.44949   2.915476  271.733029
min     1.00000   1.000000   40.000000
25%     2.75000   2.750000  133.750000
50%     4.50000   4.500000  250.000000
75%     6.25000   6.250000  525.000000
max     8.00000  10.000000  800.000000

Average quantity sold: 4.75
Maximum price: 800.0


`Answer:`
- Average quantity sold: 4.75
- Maximum price: 800.0


#### Missing Values

Use .info() to get a concise summary of df_sales.

Question: Are there any missing values?


In [7]:

df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Order_ID        8 non-null      int64 
 1   Product         8 non-null      object
 2   Category        8 non-null      object
 3   Quantity        8 non-null      int64 
 4   Price           8 non-null      int64 
 5   Store_Location  8 non-null      object
dtypes: int64(3), object(3)
memory usage: 512.0+ bytes


`Answer:` There is no missing Values

#### Min and Max Price
Find the minimum, maximum, average values of the Price column.

Question: What is the lowest price and which product has it?

In [8]:

min_price = df_sales["Price"].min()
max_price = df_sales["Price"].max()
average_price = df_sales["Price"].mean()


product_with_min_price = df_sales[df_sales["Price"] == min_price]["Product"].iloc[0]


print("Minimum price:", min_price)
print("Maximum price:", max_price)
print("Average price:", average_price)
print("Product with the lowest price:", product_with_min_price)


Minimum price: 40
Maximum price: 800
Average price: 334.375
Product with the lowest price: Lamp


`Answer:`
- Lowest price: 40
- Product with the lowest price: Lamp

## Task 3: Filtering and Slicing

#### **Filter by Category**

Filter rows where Category is "Electronics".

**Question:** How many electronic products are in the dataset?


In [9]:

electronics = df_sales[df_sales["Category"] == "Electronics"]

electronics_count = electronics.shape[0]

print("Filtered rows where Category is 'Electronics':",electronics)

print("\nNumber of electronic products:", electronics_count)


Filtered rows where Category is 'Electronics':    Order_ID     Product     Category  Quantity  Price Store_Location
0         1      Laptop  Electronics         3    800       New York
1         2  Headphones  Electronics         5    150  San Francisco
4         5     Monitor  Electronics         4    300  San Francisco
6         7  Smartphone  Electronics         6    600        Chicago

Number of electronic products: 4


`Answer:`
Number of electronic products: 4

### Filter by Price

Filter rows where Price is above 300.

**Question:** What products have a price greater than 300?

In [100]:
# CODE HERE
price = df_sales[df_sales['Price']>300]
highest_product  = price['Product']
print("the Products which are greter than 300:\n",highest_product.unique())

the Products which are greter than 300:
 ['Laptop' 'Smartphone' 'Sofa']


`Answer:` The Products which are Greater than 300 are 'Laptop','Smartphone','Sofa'. 


#### Filter by Location
Filter rows for orders from "New York".

**Question:** How many orders were made from "New York"?


In [78]:
# CODE HERE
orders = df_sales[df_sales['Store_Location'] == "New York"]
order_count = orders.shape[0]
print("Filter rows for orders from New York:",orders)
print("Number of Orders made by New York:",order_count)

Filter rows for orders from New York:    Order_ID Product     Category  Quantity  Price Store_Location
0         1  Laptop  Electronics         3    800       New York
2         3   Chair    Furniture        10     85       New York
5         6    Lamp    Furniture         7     40       New York
Number of Orders made by New York: 3


`Answer:`
Number of Orders made by New York is 3

## Indexing and Slicing

Select only the Product and Quantity columns for the first 5 rows.

**Question:** What are the names and quantities of the first five products?


In [10]:

first_five_products = df_sales.loc[:4, ["Product", "Quantity"]]

print("Product and Quantity for the first 5 rows:")
print(first_five_products)


Product and Quantity for the first 5 rows:
      Product  Quantity
0      Laptop         3
1  Headphones         5
2       Chair        10
3        Desk         2
4     Monitor         4


`Answer:`
The names and quantities of the first five products are:

    1. Laptop: 3
    2. Headphones: 5
    3. Chair: 10
    4. Desk: 2
    5. Monitor: 4

## Advanced Filtering
Select rows where Category is "Furniture" and Quantity is more than 5.

**Question:** How many furniture items have quantities greater than 5?


In [11]:

furniture_with_high_quantity = df_sales[(df_sales["Category"] == "Furniture") & (df_sales["Quantity"] > 5)]


furniture_count = furniture_with_high_quantity.shape[0]


print("Furniture items with quantities greater than 5:\n",furniture_with_high_quantity)

print("\nNumber of furniture items with quantities greater than 5:",furniture_count)



Furniture items with quantities greater than 5:
    Order_ID Product   Category  Quantity  Price Store_Location
2         3   Chair  Furniture        10     85       New York
5         6    Lamp  Furniture         7     40       New York

Number of furniture items with quantities greater than 5: 2


`Answer:` Number of furniture items with quantities greater than 5: 2

## Task 4: Adding and Modifying Columns

#### Add Total Price Column

Add a new column called Total_Price, calculated as Quantity * Price.

**Question:** What is the total price for each product?



In [12]:

df_sales["Total_Price"] = df_sales["Quantity"] * df_sales["Price"]

print("Updated DataFrame with Total_Price column:")
print(df_sales[["Product", "Total_Price"]])


Updated DataFrame with Total_Price column:
      Product  Total_Price
0      Laptop         2400
1  Headphones          750
2       Chair          850
3        Desk          400
4     Monitor         1200
5        Lamp          280
6  Smartphone         3600
7        Sofa          500


`Answer:`
The total price for each product is:

    1. Laptop: 2400
    2. Headphones: 750
    3. Chair: 850
    4. Desk: 400
    5. Monitor: 1200
    6. Lamp: 280
    7. Smartphone: 3600
    8. Sofa: 500

#### Add Price Category


Add a new column called Price_Category, labeling products as "High" if Price > 300 and "Low" otherwise.

**Question:** How many products fall into each price category?

In [13]:


df_sales["Price_Category"] = ["High" if price > 300 else "Low" for price in df_sales["Price"]]


price_category_counts = df_sales['Price_Category'].value_counts()

print("Updated DataFrame with Price_Category column:")
print(df_sales[["Product", "Price", "Price_Category"]])

print("\nNumber of products in each price category:")
price_category_counts


Updated DataFrame with Price_Category column:
      Product  Price Price_Category
0      Laptop    800           High
1  Headphones    150            Low
2       Chair     85            Low
3        Desk    200            Low
4     Monitor    300            Low
5        Lamp     40            Low
6  Smartphone    600           High
7        Sofa    500           High

Number of products in each price category:


Price_Category
Low     5
High    3
Name: count, dtype: int64

`Answer:`
- Low price products: 5
- High price products: 3






Task 5: Merging and Concatenation
Create a DataFrame df_discounts with the following data:


| Category           | Discount_Percentage |
|--------------------|---------------------|
| Electronics        | 10                  |
| Furniture          | 15                  |
| Clothing           | 20                  |
| Groceries          | 5                   |
| Toys               | 25                  |
| Books              | 30                  |
| Sports Equipment   | 12                  |
| Beauty             | 18                  |




In [14]:

import pandas as pd

data_discounts = {
    "Category": [
        "Electronics", "Furniture", "Clothing", "Groceries", 
        "Toys", "Books", "Sports Equipment", "Beauty"
    ],
    "Discount_Percentage": [10, 15, 20, 5, 25, 30, 12, 18]
}

df_discounts = pd.DataFrame(data_discounts)

print("Discounts DataFrame:")
print(df_discounts)



Discounts DataFrame:
           Category  Discount_Percentage
0       Electronics                   10
1         Furniture                   15
2          Clothing                   20
3         Groceries                    5
4              Toys                   25
5             Books                   30
6  Sports Equipment                   12
7            Beauty                   18


## Merge DataFrames
Merge df_sales and df_discounts based on the Category column using a left join. Save the result as df_combined.

**Question:** How many rows are in df_combined?


In [15]:

df_combined = pd.merge(df_sales, df_discounts, on="Category", how="left")

print("Combined DataFrame:")
print(df_combined)


num_rows = df_combined.shape[0]
print(f"\nNumber of rows in df_combined: {num_rows}")


Combined DataFrame:
   Order_ID     Product     Category  Quantity  Price Store_Location  \
0         1      Laptop  Electronics         3    800       New York   
1         2  Headphones  Electronics         5    150  San Francisco   
2         3       Chair    Furniture        10     85       New York   
3         4        Desk    Furniture         2    200        Chicago   
4         5     Monitor  Electronics         4    300  San Francisco   
5         6        Lamp    Furniture         7     40       New York   
6         7  Smartphone  Electronics         6    600        Chicago   
7         8        Sofa    Furniture         1    500  San Francisco   

   Total_Price Price_Category  Discount_Percentage  
0         2400           High                   10  
1          750            Low                   10  
2          850            Low                   15  
3          400            Low                   15  
4         1200            Low                   10  
5          28

`Answer:` Number of rows in df_combined 8

### Calculate Discounted Price
Add a new column called Discounted_Price in df_combined, using the formula: **Discounted_Price = Total_Price * (1 - Discount_Percentage / 100).**

In [16]:

df_combined['Discounted_Price'] = df_combined['Total_Price'] * (1 - df_combined['Discount_Percentage'] / 100)

print("DataFrame with Discounted_Price:")
print(df_combined)


DataFrame with Discounted_Price:
   Order_ID     Product     Category  Quantity  Price Store_Location  \
0         1      Laptop  Electronics         3    800       New York   
1         2  Headphones  Electronics         5    150  San Francisco   
2         3       Chair    Furniture        10     85       New York   
3         4        Desk    Furniture         2    200        Chicago   
4         5     Monitor  Electronics         4    300  San Francisco   
5         6        Lamp    Furniture         7     40       New York   
6         7  Smartphone  Electronics         6    600        Chicago   
7         8        Sofa    Furniture         1    500  San Francisco   

   Total_Price Price_Category  Discount_Percentage  Discounted_Price  
0         2400           High                   10            2160.0  
1          750            Low                   10             675.0  
2          850            Low                   15             722.5  
3          400            Low     


Question: What is the discounted price for each product?

Concatenate DataFrames
Concatenate df_sales with a new DataFrame df_extra_sales:

| Order_ID | Product     | Category    | Quantity | Price | Store_Location |
|----------|-------------|-------------|----------|-------|----------------|
| 1        | Laptop      | Electronics | 3        | 800   | New York       |
| 2        | Headphones  | Electronics | 5        | 150   | San Francisco  |
| 3        | Chair       | Furniture   | 10       | 85    | New York       |
| 4        | Desk        | Furniture   | 2        | 200   | Chicago        |
| 5        | Monitor     | Electronics | 4        | 300   | San Francisco  |
| 6        | Lamp        | Furniture   | 7        | 40    | New York       |
| 7        | Smartphone  | Electronics | 6        | 600   | Chicago        |
| 8        | Sofa        | Furniture   | 1        | 500   | San Francisco  |
| 9        | Table       | Furniture   | 2        | 150   | Chicago        |
| 10       | Speaker     | Electronics | 3        | 200   | San Francisco  |


Question: After concatenation, how many total rows are in the updated DataFrame?



In [17]:


df_extra_sales = pd.DataFrame({
    'Order_ID': [9, 10],
    'Product': ['Table', 'Speaker'],
    'Category': ['Furniture', 'Electronics'],
    'Quantity': [2, 3],
    'Price': [150, 200],
    'Store_Location': ['Chicago', 'San Francisco']
})

df_updated = pd.concat([df_sales, df_extra_sales], ignore_index=True)


total_rows = df_updated.shape[0]
print(f"Total number of rows after concatenation: {total_rows}")


Total number of rows after concatenation: 10


`Answer:`Total number of rows after concatenation is 10

# Task 6: Grouping and Aggregation

#### Group by Category


Group by Category to find the total quantity sold and average price per category.

**Question:** Which category sold the highest quantity overall?



In [18]:


category_summary = df_sales.groupby('Category').agg(total_quantity=('Quantity', 'sum'),average_price=('Price', 'mean')).reset_index()

highest_quantity_category = category_summary.loc[category_summary['total_quantity'].idxmax()]
print(f"The category that sold the highest quantity overall is: {highest_quantity_category['Category']}")


The category that sold the highest quantity overall is: Furniture


#### Group by Store Location
Group by Store_Location to find the total quantity sold and total Total_Price at each location.

**Question:** Which store location generated the highest revenue?



In [20]:


store_summary = df_sales.groupby('Store_Location').agg(
    total_quantity=('Quantity', 'sum'),
    total_revenue=('Total_Price', 'sum')
).reset_index()


highest_revenue_store = store_summary.loc[store_summary['total_revenue'].idxmax()]
print(f"The store location that generated the highest revenue is: {highest_revenue_store['Store_Location']}")


The store location that generated the highest revenue is: Chicago


`Answer:` The store location that generated the highest revenue is: Chicago

### Discounted Revenue Calculation
Using df_combined, group by Category and calculate the average Discounted_Price and the total revenue after discounts.

**Question:** What is the total discounted revenue for the "Electronics" category?

In [21]:


category_discounted_revenue = df_combined.groupby('Category').agg(
    total_discounted_revenue=('Discounted_Price', 'sum')
).reset_index()


electronics_revenue = category_discounted_revenue.loc[category_discounted_revenue['Category'] == 'Electronics']
print(f"The total discounted revenue for the 'Electronics' category is: {electronics_revenue['total_discounted_revenue'].values[0]}")


The total discounted revenue for the 'Electronics' category is: 7155.0


`Answer:` The total discounted revenue for the 'Electronics' category is 7155.0